File: web-app/src/repositories/FxRepository.ts
import { FxService } from '@/services/FxService';
import { LruCache } from '@/utils/LruCache';
import { DAY_MS } from '../../../packages/core/net';

/**
 * Repository providing cached access to FX rates via FxService.
 */
export class FxRepository {
  private svc: FxService;
  private cache = new LruCache<string, number>(32);

  constructor(service?: FxService) {
    this.svc = service ?? new FxService();
  }

  /**
   * Get conversion rate from base to quote using a 24 h cache.
   */
  async rate(base: string, quote: string): Promise<number | null> {
    const key = `${base}_${quote}`;
    const cached = this.cache.get(key);
    if (cached !== undefined) return cached;
    const value = await this.svc.getRate(base, quote);
    if (value !== null) this.cache.put(key, value, DAY_MS);
    return value;
  }
}

File: mobile-app/lib/repositories/fx_repository.dart
import 'package:smwa_services/services.dart';
import 'package:smwa_services/src/lru_cache.dart';

/// FR-0107 – Provides cached access to FX rates via [FxService].
class FxRepository {
  final FxService _svc;
  final LruCache<String, double> _cache = LruCache(32);

  FxRepository({FxService? service}) : _svc = service ?? FxService();

  /// Returns the conversion rate from [base] to [quote] using a 24h cache.
  Future<double?> rate(String base, String quote) async {
    final key = '${base}_${quote}';
    final cached = _cache.get(key);
    if (cached != null) return cached;
    final value = await _svc.getRate(base, quote);
    if (value != null) _cache.put(key, value, const Duration(hours: 24));
    return value;
  }
}
